src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.EnumerateMatches.Tests.cs

```diff
@@ -0,0 +1,237 @@
+            AssertExtensions.Throws<ArgumentNullException>("pattern", () => Regex.EnumerateMatches("input", null, RegexOptions.None, TimeSpan.FromSeconds(1)));
+
+            // Options are invalid
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("options", () => Regex.EnumerateMatches("input", "pattern", (RegexOptions)(-1)));
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("options", () => Regex.EnumerateMatches("input", "pattern", (RegexOptions)(-1), TimeSpan.FromSeconds(1)));
+
+            // 0x400 is new NonBacktracking mode that is now valid, 0x800 is still invalid
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("options", () => Regex.EnumerateMatches("input", "pattern", (RegexOptions)0x800));
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("options", () => Regex.EnumerateMatches("input", "pattern", (RegexOptions)0x800, TimeSpan.FromSeconds(1)));
+
+            // MatchTimeout is invalid
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("matchTimeout", () => Regex.EnumerateMatches("input", "pattern", RegexOptions.None, TimeSpan.Zero));
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("matchTimeout", () => Regex.EnumerateMatches("input", "pattern", RegexOptions.None, TimeSpan.Zero));
+        }
+
+        [Theory]
+        [MemberData(nameof(RegexHelpers.AvailableEngines_MemberData), MemberType = typeof(RegexHelpers))]
+        public void Enumerate_No_Match(RegexEngine engine)
+        {
+            Regex r = RegexHelpers.GetRegexAsync(engine, @"\da").GetAwaiter().GetResult();
+            int count = 0;
+            foreach(var match in r.EnumerateMatches("1A2b3c4d5e"))
+            {
+                count++;
+            }
+            Assert.Equal(0, count);
+
+        }
+
+        [Theory]
+        [MemberData(nameof(NoneCompiledBacktracking))]
+        public static void EnumerateMatches_Invalid(RegexOptions options)
+        {
+            Regex regex = new Regex("e", options);
+            Regex.ValueMatchEnumerator enumerator = regex.EnumerateMatches("dotnet");
+
+            Assert.True(ThrowsInvalidOperationException(ref enumerator));
+
+            while (enumerator.MoveNext()) ;
+            Assert.True(ThrowsInvalidOperationException(ref enumerator));
+
+            bool ThrowsInvalidOperationException(ref Regex.ValueMatchEnumerator enumerator)
+            {
+                try
+                {
+                    _ = enumerator.Current;
+                }
+                catch (InvalidOperationException)
+                {
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(RegexHelpers.AvailableEngines_MemberData), MemberType = typeof(RegexHelpers))]
+        public void EnumerateMatches_Lookahead(RegexEngine engine)
```

Comment:
Why are these special? If we're missing lookaheads in the general matches inputs, can we add it there instead?

Reply:
They are not special, and our `Match_Count_TestData` used bellow has lookaround tests as well. The reason why I added these here, is because I basically went and added a test for every single place we were testing the `Regex.Matches` method in order to make sure we have at least the same coverage to it. These two lookaround tests are mimicking two tests we have in Regex.KnownPattern.Tests.cs where we also use these patterns and inputs to test `Regex.Matches`.

Reply:
Ok, thanks.  It's fine for now.